lib/compiler/container.ts

```diff
@@ -0,0 +1,161 @@
+import { ClassDeclaration, SyntaxKind, TypeGuards } from 'ts-morph';
+import { Compiler } from '.';
+import { container } from '../tsdi';
+import { Component } from './component';
+import { UnitImpl } from './unit';
+import { findDeclarationForIdentifier } from './util';
+
+export class Container<T> {
+  public get name(): string {
+    return this.clazz.getNameOrThrow();
+  }
+
+  public get implName(): string {
+    return `${this.name}Impl`;
+  }
+
+  public get entryPoints(): { name: string; type: Component }[] {
+    const properties = this.clazz
+      .getProperties()
+      .filter(property => property.isAbstract());
+
+    return properties.map(property => {
+      const identifier = property
+        .getTypeNodeOrThrow()
+        .getFirstChildByKindOrThrow(SyntaxKind.Identifier);
+
+      const node = findDeclarationForIdentifier(identifier);
+      if (
+        TypeGuards.isClassDeclaration(node) ||
+        TypeGuards.isInterfaceDeclaration(node)
+      ) {
+        return {
+          name: property.getName(),
+          type: new Component(this, node)
+        };
+      }
+      throw new Error('Illegal node type for component: ' + node.print());
+    });
+  }
+
+  private _units?: UnitImpl[];
```

Comment:
Right now this field suppresses that units are created multiple times. I think it would be better to make units singletons like the components are (components have a singleton check in the constructor and return the singleton instance already existing).